nifi/nifi-nar-bundles/nifi-kite-bundle/nifi-kite-processors/src/main/java/org/apache/nifi/processors/kite/AvroRecordConverter.java

```diff
@@ -0,0 +1,279 @@
+
+/**
+ * Responsible for converting records of one Avro type to another. Supports
+ * syntax like "record.field" to unpack fields and will try to do simple type
+ * conversion.
+ */
+public class AvroRecordConverter {
+	private final Schema inputSchema;
+	private final Schema outputSchema;
+	private final Map<String, String> fieldMapping;
+
+	public AvroRecordConverter(Schema inputSchema, Schema outputSchema,
+			JsonNode fieldMapping) {
+		this.inputSchema = inputSchema;
+		this.outputSchema = outputSchema;
+		this.fieldMapping = getFieldMapping(fieldMapping);
+	}
+
+	/**
+	 * Converts one record to another given a input and output schema plus
+	 * explicit mappings for certain target fields.
+	 * 
+	 * @param record
+	 * @param inputSchema
+	 * @param outputSchema
+	 * @param fieldMapping
+	 * @return
+	 * @throws AvroConversionException
+	 */
+	public Record convert(Record input) throws AvroConversionException {
+		Record result = new Record(outputSchema);
+		for (Field outputField : outputSchema.getFields()) {
+			String inputFieldName = outputField.name();
+			if (fieldMapping.containsKey(outputField.name())) {
+				inputFieldName = fieldMapping.get(outputField.name());
+			}
+
+			List<String> fieldParts = Lists.newArrayList(inputFieldName
+					.split("\\."));
+			Record current = input;
+			while (fieldParts.size() > 1) {
+				// Step into the nested records as far as needed.
+				current = (Record) current.get(fieldParts.remove(0));
+			}
+
+			// Current should now be in the right place to read the record.
+			Field f = getFieldForName(inputFieldName, inputSchema);
+			Object content = getContentForName(input, inputFieldName,
+					input.getSchema());
+			result.put(outputField.name(),
+					convertData(content, f.schema(), outputField.schema()));
+		}
+		return result;
+	}
+
+	/**
+	 * @return the inputSchema
+	 */
+	public Schema getInputSchema() {
+		return inputSchema;
+	}
+
+	/**
+	 * @return the outputSchema
+	 */
+	public Schema getOutputSchema() {
+		return outputSchema;
+	}
+
+	/**
+	 * Converts the data from one schema to another. If the types are the same,
+	 * no change will be made, but simple conversions will be attempted for
+	 * other types.
+	 * 
+	 * @param content
+	 * @param inputSchema
+	 * @param outputSchema
+	 * @return
+	 * @throws AvroConversionException
+	 */
+	private Object convertData(Object content, Schema inputSchema,
+			Schema outputSchema) throws AvroConversionException {
+		if (content == null) {
+			// No conversion can happen here.
+			return null;
+		}
+
+		Schema nonNillInput = undoNillableSchema(inputSchema);
+		Schema nonNillOutput = undoNillableSchema(outputSchema);
+		if (nonNillInput.getType().equals(nonNillOutput.getType())) {
+			return content;
+		} else {
+			switch (nonNillOutput.getType()) {
+			case STRING:
+				// This is the easiest conversion case. Converting to string we
+				// assume
+				// that String.valueOf knows what to do.
+				return String.valueOf(content);
```

Comment:
What's the value of using `String.valueOf` here? I use that to avoid `NullPointerException`, but you've already validated that content is not `null`, which is the right way to go because null is converted to the string `"null"`.